jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java

```diff
@@ -58,19 +57,27 @@
  */
 public class MultiPartFormInputStream
 {
+    private enum State
+    {
+        UNPARSED,
+        PARSING,
+        PARSED,
+        CLOSING,
+        CLOSED
+    }
+
```

Comment:
Closing and closed aren't terms that really apply here, its whether the content has been deleted or not. Can I suggest the following states, all of which except for PARSING are terminal states:

UNPARSED - before any methods are called

PARSING - first call to parse() but before all content successfully read

PARSED - all content successfully read

ERRORED - problem reading content

DELETED - first call to deleteParts()

The only situation we have to guard against is deleteParts() being called while parse() is still happening. It is not possible for 2 threads to be in parse() nor 2 threads to be in deleteParts(), so this makes the state transitions a little simpler. I'm not even sure that PARSING is necessary, maybe the parse() method could move straight to PARSED?

I think you should put in javadoc a state transition diagram so it is very clear what is happening.

Reply:
I have added a state diagram and an explanation in the javadoc.

Reply:
That's good, much easier to follow the state transitions.